src/library/scala/util/Random.scala

```diff
@@ -70,6 +70,16 @@ class Random(val self: java.util.Random) extends AnyRef with Serializable {
    */
   def nextLong(): Long = self.nextLong()
 
+  def nextLong(n: Long): Long = {
+    if (n <= 0) throw new java.lang.IllegalArgumentException("bound must be positive")
+    if (n <= Int.MaxValue) nextInt(n.toInt).toLong
+    else {
+      val RAND_MAX = Long.MaxValue     // 2^63-1
+      val x = nextLong() >>> 1         // [0, 2^63), strip the sign bit
+      if (x < (RAND_MAX - (RAND_MAX % n))) x % n else nextLong(n)
```

Comment:
The apparently tail-recursive call here is not actually tail-recursive, because `nextLong` is neither final nor private. Moreover, it will re-do the first two checks every time for no reason. Consider an explicit loop (which also avoids redoing `RAND_MAX % n` on every iteration:
```scala
else {
  val RandMax = Long.MaxValue
  val bound = RandMax - (RandMax % n)
  var x = nextLong() >>> 1
  while (x >= bound)
    x = nextLong() >>> 1
  x % n
}
```